### PR TITLE
add basic auth and storage api

### PR DESCRIPTION
**Pull request by kyipho**
As titled, adding some basic functions to:
- Sign up (no email confirmations needed for now so just using random emails will work fine)
- Sign in
- Sign out

As you'll notice, they're quite simplistic, so feel free to let me know if it's easier to do without them. Same goes the other APIs - I'll be working the others the next few days, but if any of the client ones listed at https://supabase.io/docs/reference/javascript/supabase-client are good enough to be used directly, I can exclude them. Happy to minimise any unnecessary overhead :)

Following up from my email about the new server instance, only authenticated users should be able to upload objects, so you can try something like:

```
// should fail
const file = await supabase.from('public').upload('....glb')

// should work
const {user} = await signUp('beffjezos@amzn.com', '123456')
const file = await supabase.from('public').upload('....glb')
```
Feel free to create multiple users / upload objects etc., they're easy to clean up if needed.

Also, if you'd like to look around at the server control panel / dashboard, you can sign up for an account at https://app.supabase.io/ and then I'll be able to add you to it!

### CODE REVIEW

**Comment by kyipho** [comment]
Added a second file `storage.js` to get started on the Storage API. So far only has upload and download functions, will add the 'list all objects' function soon. Please use these functions to upload/download, because I added an additional condition to the storage bucket: **users can only upload files into their own folder within the bucket**, so it needs to be written in a specific way (can be seen in `storage.js`)

File: src/api/auth.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+/* Basic auth API for Supabase. */
+
+/**
+ * Sign up with email and password. Temporarily removed
+ * requirement for email confirmations upon sign-up.
+ */
+const signUp = async (email, password) => {
+  const { user, session, error } = await supabase.auth.signUp({
```
Comment (karsheng) [inline src/api/auth.js:8]: this assumes that the supabase instance is initialized globally right?

**Comment by karsheng** [comment]
Can you take a look at this library and let me know what you think? It's one of the community libraries listed on Supabase.
https://react-supabase.vercel.app/

File: src/api/auth.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
+/* Basic auth API for Supabase. */
+
+/**
+ * Sign up with email and password. Temporarily removed
```
Comment (karsheng) [inline src/api/auth.js:4]: It still required email confirmation (by clicking on the link sent to the email address) when I was testing sign up.

File: src/api/auth.js (hunk 1 above)
Comment (kyipho) [reply src/api/auth.js:8]: Yes, quite flexible regarding the actual approach but it'd probably be something like defining a separate file `supabase.js` where the init is done, and it exports a `supabase` object. Then other files, including the API, can import it for use.

File: src/api/auth.js (hunk 2 above)
Comment (kyipho) [reply src/api/auth.js:4]: Oh that's weird, and you'd already switched to the EU server?

**Comment by kyipho** [comment]
> Can you take a look at this library and let me know what you think? It's one of the community libraries listed on Supabase.
https://react-supabase.vercel.app/

My knowledge on front-end stuff is quite limited, but it looks like it's a package to introduce some hooks for convenience? I think we don't have to explore it unless the current approach we're using is difficult for you to work with. Some reasons:
- It introduces additional coupling between the front-end with the back-end - the code to access the back-end resides directly within the React components. On the other hand, if we go with an API-ish approach, the front-end can just always rely on a `upload/downloadObject` function, and the back-end can always be developed independently, even if someday we don't want to use Supabase anymore.
- From the link, it doesn't look like they're done with the Storage functions, which will be a core part of what we need.

Feel free to disagree though. If the hooks will make a big difference to your workflow, happy to explore it too.

**Comment by kyipho** [comment]
Just added a new file `models.js` that implements some functions for working with models in the database. Key points:

### JS object structure for a model

Under `insertModel()` in `models.js`, the comments show a suggested structure for the model JS object (calling it JS object to distinguish it from a GLB/GLTF object) that differs slightly from what we discussed previously. While we talked about having one JS object for the model, another for the colours, another for the material etc., I'd suggest having it all in the same JS object like this:
```js
const model = {
  laces: {
    objects: [lace0.glb, lace1.gltf,...],
    colors: [hex0, hex1, ...],
    materials: [mesh0, mesh1, ...]
  },
  soles: {
    objects: [...],
    colors: [...],
    materials: [...],
  },
  ...
}
```
This way, we don't have to store different things in different places - a single db call will return all the information required. Let me know what you think, and whether any of the names / terminology should be changed.

### Should we use TS?

Given how a JS object for a model might be quite complex, do you think we should switch to TypeScript and enforce some types? This might make it easier to document and debug any operations done on a model. For example, something like:

```ts
interface TModelProps {
  objects: Array<string>;
  colors: Array<string>;
  materials: Array<string>
}

interface TModel {
  [key: string]: TModelProps;
}

// ensures correct structure
const model: TModel = {
  colors: [...],  // accepted
  colours: [...],  // rejected
  items: [...], // also rejected 
}
```

---

We could get on an additional call to talk about these things - the data model will require some iteration and discussion before we find something suitable.

File: src/api/auth.js (hunk 2 above)
Comment (karsheng) [reply src/api/auth.js:4]: I was still using the old server keys. Switching to EU's worked as expected.

**Comment by karsheng** [comment]
The data model you proposed looks good to me. Let's move forward this way and see how it goes.

I'm not familiar with TS. Will need an intro lesson from you :)

**Comment by kyipho** [comment]
Haha sure, I'm figuring it out myself but we can go through that on a call. The type-safety would be nice, but I'm not sure if it'll be worth the additional overhead of changing everything to a different language and set of tools. Let's talk more about it.

Also, to prevent this PR from getting too large, could you also review the changes here as well? If it's ok, then we can merge into `main` and then branch off with new stuff again.

**Pull request merged**